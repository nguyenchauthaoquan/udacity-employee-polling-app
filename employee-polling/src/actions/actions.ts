import {
    AUTHENTICATE_USER,
    GET_QUESTIONS,
    GET_USERS,
    LOGOUT,
    SAVE_ANSWERS_QUESTION,
    SAVE_QUESTION, SAVE_USER_ANSWERS, SAVE_USER_QUESTION
} from "../constants/constants.ts";
import {Action} from "../models/components/action.ts";
import {Dispatch} from "redux";
import {_getQuestions, _getUsers, _saveQuestion, _saveQuestionAnswer} from "../utils/_DATA.ts";
import User from "../models/data/user.ts";
import {reducer} from "../reducers/reducer.ts";
import Question from "../models/data/question.ts";

export const authenticateUserAction = (payload: object): Action => {
    return {
        type: AUTHENTICATE_USER,
        payload
    }
}

export const logoutAction = (): Action => {
    return {
        type: LOGOUT,
    }
}

export const getUsersAction = (payload: object): Action => {
    return {
        type: GET_USERS,
        payload
    }
}

export const getQuestions = (payload: object) => {
    return {
        type: GET_QUESTIONS,
        payload
    }
}

export const saveUserQuestionAction = (payload: object) => {
    return {
        type: SAVE_USER_QUESTION,
        payload
    }
}

export const saveUserAnswerAction = (payload: object) => {
    console.log(payload)
    return {
        type: SAVE_USER_ANSWERS,
        payload,
    }
}

export const saveQuestionAction = (payload: Question) => {
    return {
        type: SAVE_QUESTION,
        payload,
    }
}

export const saveAnswerQuestionAction = (payload: object) => {
    return {
        type: SAVE_ANSWERS_QUESTION,
        payload,
    }
}

export const getAppData = (): (dispatch: Dispatch<Action>) => Promise<void> => {
    return async (dispatch: Dispatch<Action>) => {
        const [users, questions] = await Promise.all([_getUsers(), _getQuestions()]);
        const {users: users_1, questions: questions_1} = ({
            users,
            questions
        });
        dispatch(getQuestions(questions_1));
        dispatch(getUsersAction(users_1));
    }
}

export function handleLogin(username: string, password: string) {
    return (dispatch: Dispatch<Action>, getState: () => ReturnType<typeof reducer>) => {
        const { users} = getState();

        const user = Object.values(users as User).find((user: User) => user.id === username && user.password === password);

        if (user) {
            return dispatch(authenticateUserAction(user));
        }
    };
}

export function handleLogout(): (dispatch: Dispatch<Action>) => Action {
    return (dispatch: Dispatch<Action>) => {
        return dispatch(logoutAction());
    };
}

export function handleSaveQuestion(optionOne: string, optionTwo: string) {
    return async (dispatch: Dispatch<Action>, getState: () => ReturnType<typeof reducer>) => {
        const {authenticatedUser} = getState();
        const user = authenticatedUser as User
        const question = await _saveQuestion({
            optionOneText: optionOne,
            optionTwoText: optionTwo,
            author: user
        }) as Question;

        dispatch(saveQuestionAction(question));
        dispatch(saveUserQuestionAction({
            author: question.author,
            id: question.id
        }));

    }
}

export function handleSaveAnswerQuestion(questionId: string, answer: string) {
    return (dispatch: Dispatch<Action>, getState: () => ReturnType<typeof reducer>) => {
        const {authenticatedUser} = getState();
        const user: User = authenticatedUser as User;

        _saveQuestionAnswer({
            authedUser: user.id,
            qid: questionId,
            answer: answer
        }).then(() => {
            dispatch(saveAnswerQuestionAction({
                author: user.id,
                qid: questionId,
                answer: answer
            }));
            dispatch(saveUserAnswerAction({
                authedUser: user.id,
                qid: questionId,
                answer: answer
            }))
        })
    }
}