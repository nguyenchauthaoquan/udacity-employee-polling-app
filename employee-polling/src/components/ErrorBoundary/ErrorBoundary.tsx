import { Component } from "react";
import {ErrorBoundaryProps} from "../../models/components/props.ts";
import {ErrorBoundaryState} from "../../models/components/state.ts";

export default class ErrorBoundary extends Component<ErrorBoundaryProps, ErrorBoundaryState> {
    constructor(props: ErrorBoundaryProps) {
        super(props);

        this.state = {
            hasError: false,
            errorMessage: ""
        }
    }

    static getDerivedStateFromError(error: Error): ErrorBoundaryState {
        // Update state so the next render will show the fallback UI.
        console.error(error)
        return { errorMessage: error.message, hasError: true };
    }

    componentDidCatch(error: Error, errorInfo: React.ErrorInfo) {
        console.error(error, errorInfo)
        this.setState({ errorMessage: error.message, hasError: true });
    }

    render() {
        if (this.state.hasError) {
            return <h1>{this.state.errorMessage}</h1>;
        }

        return this.props.children;
    }

}