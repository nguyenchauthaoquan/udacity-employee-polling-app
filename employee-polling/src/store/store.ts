import {configureStore} from '@reduxjs/toolkit';
import {reducer} from "../reducers/reducer.ts";

export const store = configureStore(
    {
        reducer: reducer
    }
)