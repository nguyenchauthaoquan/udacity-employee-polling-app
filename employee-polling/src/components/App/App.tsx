import {FC, useEffect} from 'react'
import {AppProps} from "../../models/components/props.ts";
import Login from "../Login/Login.tsx";
import {Route, Routes} from "react-router-dom";
import LeaderBoard from "../LeaderBoard/LeaderBoard.tsx";
import Home from "../Home/Home.tsx";
import NewPoll from "../NewPoll/NewPoll.tsx";
import {connect} from "react-redux";
import {getAppData} from "../../actions/actions.ts";
import {reducer} from "../../reducers/reducer.ts";
import {ThunkDispatch} from "redux-thunk";
import {Action} from "../../models/components/action.ts";
import User from "../../models/data/user.ts";
import Layout from "../Layout/Layout.tsx";
import PollPage from "../PollPage/PollPage.tsx";

const App: FC<AppProps> = (props: AppProps) => {
    useEffect(() => {
        props.getAppData()
    })
  return (
      <Layout>
          <Routes>
              <Route path={"/"} element={<Login/>}/>
              <Route path={"home"} element={<Home/>}/>
              <Route path={"leaderboard"} element={<LeaderBoard/>}/>
              <Route path={"new"} element={<NewPoll />}/>
              <Route path={"home/question/:id"} element={<PollPage />}/>
          </Routes>
      </Layout>
  )
}

const mapStateToProps = (state: ReturnType<typeof reducer>) => {
    return {
        user: state.authenticatedUser as User
    }
}

const mapDispatchToProps = (dispatch: ThunkDispatch<ReturnType<typeof reducer>, never, Action>) => {
    return {
        getAppData: () => {
            return dispatch(getAppData())
        }
    }
}

export default connect(mapStateToProps, mapDispatchToProps)(App);
