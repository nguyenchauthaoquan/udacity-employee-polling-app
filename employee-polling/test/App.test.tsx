import App from '../src/components/App/App'
import { render } from './test-utils'
import {Provider} from "react-redux";
import {store} from "../src/store/store";
import {BrowserRouter} from "react-router-dom";
import ErrorBoundary from "../src/components/ErrorBoundary/ErrorBoundary";
import {authenticateUserAction} from "../src/actions/actions";

describe('App', () => {
    it('should initialize the app', () => {
        const renderedComponent = render(
            <ErrorBoundary>
                <Provider store={store}>
                    <BrowserRouter>
                        <App />
                    </BrowserRouter>
                </Provider>
            </ErrorBoundary>
        )
        expect(renderedComponent).toBeDefined();
        expect(renderedComponent).toMatchSnapshot()
    })
})

describe("Login Component", () => {
    it('should show Login page when not logged in', () => {
        const renderedComponent = render(
            <ErrorBoundary>
                <Provider store={store}>
                    <BrowserRouter>
                        <App />
                    </BrowserRouter>
                </Provider>
            </ErrorBoundary>
        )
        const loginHeadingElement = renderedComponent.getByTestId("login-heading");
        expect(renderedComponent).toBeDefined();
        expect(loginHeadingElement).toBeInTheDocument();
    })

})

describe("Layout Component", () => {
    it("should show navbar when logged in successfully", () => {
        store.dispatch(authenticateUserAction({
            id: "",
            password: ""
        }))

        const renderedComponent = render(
            <ErrorBoundary>
                <Provider store={store}>
                    <BrowserRouter>
                        <App />
                    </BrowserRouter>
                </Provider>
            </ErrorBoundary>
        )

        const navbarElement = renderedComponent.getByTestId("navbar");

        expect(renderedComponent).toBeDefined();
        expect(navbarElement).toBeInTheDocument();
    })
})