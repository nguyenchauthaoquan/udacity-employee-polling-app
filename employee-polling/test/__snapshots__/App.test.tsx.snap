// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`App > should initialize the app 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="container"
      >
        <div
          class="row"
        >
          <div
            class="col-md-12"
          >
            <div
              class="container"
            >
              <div
                class="row"
              >
                <div
                  class="d-flex justify-content-center col-sm-12"
                >
                  <h3>
                    Employee Polls
                  </h3>
                </div>
              </div>
              <div
                class="row"
              >
                <div
                  class="d-flex justify-content-center col-sm-12"
                >
                  <h3
                    data-testid="login-heading"
                  >
                    Login
                  </h3>
                </div>
              </div>
              <div
                class="row"
              >
                <div
                  class="col-sm-12"
                >
                  <form
                    class=""
                  >
                    <div
                      class="mb-3"
                    >
                      <label
                        class="form-label"
                        for="username"
                      >
                        User
                      </label>
                      <input
                        class="form-control"
                        data-testid="username"
                        id="username"
                        placeholder="User"
                        type="text"
                        value=""
                      />
                    </div>
                    <div
                      class="mb-3"
                    >
                      <label
                        class="form-label"
                        for="password"
                      >
                        Password
                      </label>
                      <input
                        class="form-control"
                        data-testid="password"
                        id="password"
                        placeholder="Password"
                        type="password"
                        value=""
                      />
                    </div>
                    <button
                      class="btn btn-light"
                      data-testid="submit"
                      type="submit"
                    >
                      Login
                    </button>
                  </form>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="container"
    >
      <div
        class="row"
      >
        <div
          class="col-md-12"
        >
          <div
            class="container"
          >
            <div
              class="row"
            >
              <div
                class="d-flex justify-content-center col-sm-12"
              >
                <h3>
                  Employee Polls
                </h3>
              </div>
            </div>
            <div
              class="row"
            >
              <div
                class="d-flex justify-content-center col-sm-12"
              >
                <h3
                  data-testid="login-heading"
                >
                  Login
                </h3>
              </div>
            </div>
            <div
              class="row"
            >
              <div
                class="col-sm-12"
              >
                <form
                  class=""
                >
                  <div
                    class="mb-3"
                  >
                    <label
                      class="form-label"
                      for="username"
                    >
                      User
                    </label>
                    <input
                      class="form-control"
                      data-testid="username"
                      id="username"
                      placeholder="User"
                      type="text"
                      value=""
                    />
                  </div>
                  <div
                    class="mb-3"
                  >
                    <label
                      class="form-label"
                      for="password"
                    >
                      Password
                    </label>
                    <input
                      class="form-control"
                      data-testid="password"
                      id="password"
                      placeholder="Password"
                      type="password"
                      value=""
                    />
                  </div>
                  <button
                    class="btn btn-light"
                    data-testid="submit"
                    type="submit"
                  >
                    Login
                  </button>
                </form>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
